### PR TITLE
Entrypoint vcl cleanup. Fix #78

### LINKED ISSUE #78: Default entrypointVCLFileContent issues
[Varnish by default appends its `builtin.vcl` to custom logic](https://varnish-cache.org/docs/trunk/users-guide/vcl-built-in-code.html#built-in-vcl)

Right now, default `entrypointVCLFileContent` contains several calls to return preventing `builtin.vcl` execution this could cause several issues or weird behaviors.

IMHO, `entrypointVCLFileContent` should better stick to what is needed and improve [its documentation](https://ibm.github.io/varnish-operator/vcl-configuration.html)

Issues/weirdness encountered with current `entrypointVCLFileContent`:
* **All** request are transformed to `GET`: Varnish by default only deals with `GET` & `HEAD`, see https://github.com/varnishcache/varnish-cache/blob/6.5/bin/varnishd/builtin.vcl#L65, but currently `entrypointVCLFileContent` includes an insensitive `return (hash);` preventing built-in logic to kick in. AFAIK it's not documented
* Only url is used for request hashing: This makes the config unsafe when backends varying responses upon `host` header. Imaging a PHP application listening to both `api.my.app/node/1` and `www.my.app/node/1`. AFAIK it's not documented
* [Stale objects are never used](https://github.com/IBM/varnish-operator/blob/main/pkg/varnishcluster/controller/varnishcluster_default_vcl.go#L95): AFAIK it's not documented

I'll suggest to remove unneeded returns from `entrypointVCLFileContent`, it may both limit weirdness and ease maintanibility

Best regards

### PR DESCRIPTION
**Pull request by NITEMAN**
Fix #78

### CODE REVIEW

File: pkg/varnishcluster/controller/varnishcluster_default_vcl.go [hunk 1]
```diff
@@ -40,8 +39,6 @@ sub vcl_recv {
   if (req.url ~ "health") {
     return (pass);
   }
-
-  return (hash);
 }
```
Comment (cin) [inline pkg/varnishcluster/controller/varnishcluster_default_vcl.go:42]: Have been going back and forth with one of our devops guys. He mentioned that we typically drop cookies before the default `vcl_recv` logic -- since the built-in logic won't cache anything with cookies. Thoughts on doing that here too?

```suggestion
  # Drop cookies for anything else
  unset req.http.Cookie;
}
```

Comment (NITEMAN) [reply pkg/varnishcluster/controller/varnishcluster_default_vcl.go:42]: @cin You're right, [`builtin.vcl` intentionally avoids caching request which has cookies](https://github.com/varnishcache/varnish-cache/blob/6.5/bin/varnishd/builtin.vcl#L69). As far as I can recall, this is due security reasons since a lot of webapps use cookies for auth & session handling (Drupal and Wordpress just to name two well kown).

If the back-end uses cookies for session handling and you want to cache requests anyway, the usual practice in VCL is hashing on cookie contents (well, usually on filtered cookie contents). You can see an example of [filtering](https://github.com/NITEMAN/varnish-bites/blob/master/varnish4/drupal-base.vcl#L348) and then [hashing](https://github.com/NITEMAN/varnish-bites/blob/master/varnish4/drupal-base.vcl#L455). If you don't handle auth (or variation) cookies that way you can end up serving private results from cache to anonymous users.

`builtin.vcl`, as a general purpose base, plays for safe and skip caching upon the presence of any cookies avoiding the potential security hole.

As far as I can understand, the underlying issue is that `varnish-operator` is currently widely deployed using an opinionated VCL and changes on it may _break_ existent deployments.

So it's a matter of design decision on your part:
- If `entrypointVCLFileContent` is opinionated (perfectly fair): This just should be reflected on the docs
- If `entrypointVCLFileContent` is general purpose / un-opinionated: IMHO VCL should stick to `builtin.vcl` as much as posible

Either decision is fine (we are forced to use custom VCLs anyway) and I'll adjust the PR accordingly, just let me know ;)

Comment (cin) [reply pkg/varnishcluster/controller/varnishcluster_default_vcl.go:42]: > As far as I can understand, the underlying issue is that varnish-operator is currently widely deployed using an opinionated VCL and changes on it may break existent deployments.

This is true. We've tried to be as prescriptive as possible in the operator, so it (hopefully) keeps things simple for users. However, you've pointed out a few flaws/best practices in the VCL that we should correct. Also, I'm not as concerned about busting cache bc we'll bust cache when we release a new version anyway (all the pods have to roll). We may bust cache in a more catastrophic way given the hashing change is going to alter which shards own what data. That said, we can always mitigate this on our end by overriding the VCL in specific applications where the old behavior is needed. So, I'd rather get this right for users of the operator.

Given the security issues you've pointed out, we should probably not do anything with cookies and let each deployment deal with them as needed. I think that will keep things simple as I don't see a good way to deal with all use cases here. We could maybe allow templating around this, but that seems like overkill. Thoughts?

TLDR: I'd like to move forward with your PR. :) Let's hammer out how we want to treat cookies and see if there's anything else we should add to the docs to help clarify things for end users.

Comment (NITEMAN) [reply pkg/varnishcluster/controller/varnishcluster_default_vcl.go:42]: Well, VCL it's a bit tricky and I'm almost a newcomer (~2 years) to k8s... but I'll try my best

Thinking aloud...

As I don't know how to make fields available to template engine... although it may be possible to signal/choose some _usual_ behaviors mixing env vars and conditional logic.

One possible drawback is, since it is mandatory in VCL to use any declarations (i.e: defined subs must be called somewhere), that it might make the template (and associated doc) look a bit obscure / harder to follow.

Quick&Dirty implementation for configurable cookie removal:
```diff
diff --git a/pkg/varnishcluster/controller/varnishcluster_default_vcl.go b/pkg/varnishcluster/controller/varnishcluster_default_vcl.go
index 5316f68..bb01cd0 100644
--- a/pkg/varnishcluster/controller/varnishcluster_default_vcl.go
+++ b/pkg/varnishcluster/controller/varnishcluster_default_vcl.go
@@ -39,6 +39,8 @@ sub vcl_recv {
   if (req.url ~ "health") {
     return (pass);
   }
+  // Deal with cookies
+  call sub vcl_req_cookies;
 }
 
 sub vcl_synth {
@@ -133,6 +135,14 @@ backend dummy {
 }
 {{- end }}
 
+// Custom subs
+sub vcl_req_cookies {
+  if ( req.http.Cookie
+    && std.getenv("remove_cookies") ~ "(?i)(true|1)" ) {
+      unset req.http.Cookie;
+  }
+}
+
 sub init_backends {
   // The line below is generated and creates a variable that is used to build custom logic
   // when the user configured the backends incorrectly. E.g. return a custom error page that indicates the issue.
```

`std` & `var` give needed flexibility to some extent, but again its a matter of deciding what the behavior should be.

I'll adjust the PR accordingly, just let me know ;)

Comment (cin) [reply pkg/varnishcluster/controller/varnishcluster_default_vcl.go:42]: @NITEMAN sorry for the delay. I had to take a sick day yesterday.  While I think the idea above is good, I think we should let users handle cookies by overriding the VCL. Given the complexities of what they may want to do with cookies and how they may be used as part of the hash, I think this is the best approach and assumes less about the application. However, I'm open to suggestions.

Comment (cin) [reply pkg/varnishcluster/controller/varnishcluster_default_vcl.go:42]: Just let the tests run. Seeing an error [here](https://github.com/IBM/varnish-operator/actions/runs/3144805671/jobs/5127161748). I haven't had a chance to investigate yet. Just pointing it out.

Comment (cin) [reply pkg/varnishcluster/controller/varnishcluster_default_vcl.go:42]: Was just a glitch in the matrix apparently. Noticed that it was only impacting one version of k8s, so I just re-ran the test and it worked.

Comment (NITEMAN) [reply pkg/varnishcluster/controller/varnishcluster_default_vcl.go:42]: @cin There's no hurry, I hope you are better.

IMHO any further work on cookies would better fit in separate MRs, covered with some tests if possible. Future community request will probably provide an invaluable guidance on what is needed / demanded.

For me this is ready to be merged and fully address all my concerns in #78

Comment (cin) [reply pkg/varnishcluster/controller/varnishcluster_default_vcl.go:42]: 100% agree @NITEMAN.

**Review by cin — APPROVED** [review]
LGTM!
